### PR TITLE
avr: initial implementation for PWM

### PR DESCRIPTION
**Pull request by deadprogram**
This is a initial implementation of PWM for the AVR on the Arduino Uno. It mimics the pattern established by `GPIO` but with appropriate changes for the setting of PWM on pins that support it.

There is also an included example program, that has been tested on an Arduino Uno with an RGB LED connected.

### CODE REVIEW

File: src/machine/machine_avr.go [hunk 1]
```diff
@@ -58,3 +58,93 @@ func (p GPIO) Get() bool {
...
+
+	CS00 = 0
+	CS01 = 1
+	CS11 = 1
+	CS22 = 2
+
+	WGM10 = 0
+	WGM11 = 1
+	WGM20 = 0
+	WGM21 = 1
```
Comment (aykevl) [inline src/machine/machine_avr.go]: These values should ideally not be hardcoded here, but automatically generated from the .atdf file of this device. The source is located at `lib/avr/packs/atmega/ATmega328P.atdf` and the script to generate the bitfields is `tools/gen-device-avr.py`.
It will probably require some guesswork, trial and error to get them right, but it will avoid having to hardcode constants in the source here.

Also, I generally avoid specifying literal bit numbers as is custom for the AVRs because it leads to a lot of code like `(1 << foo)`, see for example the code in `src/runtime/runtime_avr.go` that avoids it. I would instead prefer the actual values so you can use them directly (e.g. simply `*avr.TCCR0B |= avr.CS01`). Unless, of course, you know a good reason not to do it like this.

File: src/machine/machine_avr.go [hunk 2]
```diff
@@ -58,3 +58,93 @@ func (p GPIO) Get() bool {
...
+)
+
+// InitPWM initializes the registers needed for PWM.
+func InitPWM() {
+	// use waveform generation
+	*avr.TCCR0A |= avr.TCCR0A_WGM0
+
+	// set timer 0 prescale factor to 64
+	*avr.TCCR0B |= 1 << CS01
+	*avr.TCCR0B |= 1 << CS00
```
Comment (aykevl) [inline src/machine/machine_avr.go]: Merging these two lines probably reduces code size, as writing to `*avr.TCCR0B` cannot be optimized (the pointer is marked volatile). This should be more compact and equivalent (but see the comment above regarding bit numbers):

```go
	*avr.TCCR0B |= (1 << CS01) | (1 << CS00)
```

File: src/machine/machine_avr.go [hunk 3]
```diff
@@ -58,3 +58,93 @@ func (p GPIO) Get() bool {
...
+// Configure configures a PWM pin for output.
+func (pwm PWM) Configure() {
+	if pwm.Pin < 8 {
+		*avr.DDRD |= 1 << pwm.Pin
+	} else {
+		*avr.DDRB |= 1 << (pwm.Pin - 8)
+	}
+}
+
+// Set sets the needed register values to turn on the duty cycle for a PWM pin.
```
Comment (aykevl) [inline src/machine/machine_avr.go]: While correct, the whole point of this function is setting register values. Perhaps this description can be shortened to:

> Set turns on the duty cycle for a PWM pin.

Also, it is perhaps best to describe in what units this is - something like "0 means 0% on and 255 means 100% on" (if that is the case, I seem to remember some issue with the AVR in which PWM would not normally reach 100% when fully enabled).

At the same time, I think we should use 16-bit values here. This should be optimized to 8-bit values when calling PWM.Set(value) with a constant value known at compile time, but gives greater flexibility for platforms that do support more than 8 bits of PWM.

File: src/machine/machine_avr.go [hunk 4]
```diff
@@ -58,3 +58,93 @@ func (p GPIO) Get() bool {
...
+		// connect pwm to pin on timer 1, channel B
+		*avr.TCCR1A |= 1 << COM1B1
+		// this is a 16-bit value, but we only currently allow the low order bits to be set
+		*avr.OCR1BL = avr.RegValue(value) // set pwm duty
+	case 11:
+		// connect pwm to pin on timer 2, channel A
+		*avr.TCCR2A |= 1 << COM2A1
+		*avr.OCR2A = avr.RegValue(value) // set pwm duty
+	default:
+		// TODO: handle invalid pin for PWM on Arduino
```
Comment (aykevl) [inline src/machine/machine_avr.go]: I think it's best to specify what should happen here, and document it at the function documentation. Either not doing anything, or panicking (both is fine to me). Panicking should be fine in case the pin number is known at compile time, as this whole switch will be optimized away and thus also the call to `panic()`.

File: src/machine/machine.go [hunk 5]
```diff
@@ -15,3 +15,7 @@ func (p GPIO) High() {
 func (p GPIO) Low() {
 	p.Set(false)
 }
+
+type PWM struct {
+	Pin uint8
```
Comment (aykevl) [inline src/machine/machine.go]: Is there a reason not to re-use the `GPIO` object? Perhaps it can be renamed to `Pin` if it supports more than just setting pins.

File: src/machine/machine_avr.go [hunk 6]
```diff
@@ -58,3 +58,93 @@ func (p GPIO) Get() bool {
...
+		*avr.DDRD |= 1 << pwm.Pin
+	} else {
+		*avr.DDRB |= 1 << (pwm.Pin - 8)
+	}
+}
+
+// Set sets the needed register values to turn on the duty cycle for a PWM pin.
+func (pwm PWM) Set(value uint8) {
+	switch pwm.Pin {
+	case 3:
```
Comment (aykevl) [inline src/machine/machine_avr.go]: These numbers should be constants - something like:
```go
const (
    PWM0 = 3
    PWM1 = 5
    // etc
)
```
Constants are also easier to share with other chips, e.g. the nrf52 also has PWM but on different pins.

Comment (deadprogram) [reply src/machine/machine_avr.go]: The basic philosophy I have tried try to follow is to use whatever is silkscreened on the circuit board. 

Since the Arduino labels the board itself with the pin numbers  "3", "5", etc. I figured it would be a lot easier for users to determine how to write the code that does what they expect. Creating a separate naming/indexing system aka PWM0 etc will require another lookup of the docs for a dev to figure out that PWM0 means pin 3 on Arduino, and some other pin on nrf52.

Perhaps it would be better to have separate example for PWM on nrf52 and eventually any other platforms that are added that support a PWM interface.

File: src/machine/machine.go (hunk 5 above)
Comment (deadprogram) [reply src/machine/machine.go]: Well, actually, I was windering if perhaps it might be a good idea to create a GPIO interface, and separately a PWM interface, and then have a specific machine implement whatever interface(s) for a pin on that platform. This would help to ensure some consistency across platforms. WHat do you think?

File: src/machine/machine_avr.go (hunk 4 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: In that case a panic() seems like it might be the best here for compile time check. 

How should it handle the cases someone tries to specify at runtime which pin to use? For example to write a Firmata-like implementation that is controlled by an external caller using serial interface, or wireless interface or something.

File: src/machine/machine_avr.go (hunk 2 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: Merging into single line seems like a good idea.

File: src/machine/machine_avr.go (hunk 1 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: I dug thru the atdf files for a while before I added those constants to `machine_avr.go`. None of them appear to be values that match the section that seem like they should belong to. For example `TCCR0B` we need to set to 0x03. Using your suggested syntax:

```go
*avr.TCCR0B |= (1 << CS01) | (1 << CS00)
```

However the atdf file does not have a value that corresponds to 0x03 for the `TCCR0B` register:

```xml
<register caption="Timer/Counter Control Register B" name="TCCR0B" offset="0x45" size="1">
    <bitfield caption="Force Output Compare A" mask="0x80" name="FOC0A"/>
    <bitfield caption="Force Output Compare B" mask="0x40" name="FOC0B"/>
    <bitfield caption="" mask="0x08" name="WGM02"/>
    <bitfield caption="Clock Select" mask="0x07" name="CS0" values="CLK_SEL_3BIT_EXT"/>
</register>
```

I found these values by digging thru the Arduino AVR source code https://github.com/arduino/ArduinoCore-avr/blob/master/cores/arduino/wiring_analog.c and sone AVR docs such as https://sites.google.com/site/qeewiki/books/avr-guide/pwm-on-the-atmega328

Anyhow, there does not seem to be a way to parse the values if they are not present in the atdf file. What do you suggest?

Comment (aykevl) [reply src/machine/machine_avr.go]: I *think* the extra number at the end is the 0-indexed bit number. So for example, there is `COM2A` with value 0xC0 in the .atdf file, which is 11000000 in binary. So there are two bits in this part of the register. This matches with the two bits of the register that are specified in the C header files: `COM2A0` and `COM2A1`.
The same goes for `CS0`, which has 3 bits (0x07) and has values `CS00`, `CS01` and `CS02` (with values 0, 1, and 2).

File: src/machine/machine_avr.go (hunk 4 above)
Comment (aykevl) [reply src/machine/machine_avr.go]: When it is specified at runtime (or LLVM can't deduce it - it isn't magical) it will be a regular switch, like before. And that means the panic will be compiled in the binary, which you may or may not want. I don't have a strong preference for either: having a panic here makes finding bugs easier in some cases but may also increase firmware size.

File: src/machine/machine.go (hunk 5 above)
Comment (aykevl) [reply src/machine/machine.go]: I don't think I understand the need for interfaces here? It's nice to have a similar set of methods as other APIs (for duck typing) but otherwise I don't see the need. TinyGo programs are compiled for one specific chip so can simply use a `machine.GPIO` (for example) without having to care too much how that is implemented. This doesn't need an interface, just a similar set of methods across platforms.

Aside from that, I'm a little bit cautious with interfaces, as they can cause code bloat. I've done some optimizations that reduce the overhead by a large margin but still they aren't free so should be used only when necessary.

File: src/machine/machine_avr.go (hunk 6 above)
Comment (aykevl) [reply src/machine/machine_avr.go]: Ah, I mixed ADC and PWM - ADC is numbered differently. Yes, then it makes sense to keep these numbers. It's a bit unfortunate that it is hard to share example code across platforms but so be it.

I also don't really like how these numbers are hardcoded for all AVR chips (in machine_avr.go) but let's keep it as is for now - it can be changed when a different AVR comes along with a different PWM numbering system.

File: src/examples/pwm/pwm.go [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	for {
+		rc = cycleColor(rc)
+		gc = cycleColor(gc)
+		bc = cycleColor(bc)
+
+		red.Set(rc)
+		green.Set(gc)
+		blue.Set(bc)
+
+		runtime.Sleep(runtime.Millisecond * 500)
```
Comment (aykevl) [inline src/examples/pwm/pwm.go]: I just did an update after which you can directly use the "time" package: `time.Sleep(time.Millisecond * 500)`. See: 2a20c0c7f0c13c9b5c81422ea1c24ddd4eb765e7. This means the above code won't compile after a merge.

File: src/machine/machine_avr.go (hunk 1 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: Updated `gen-device-avr.py` based on your feedback.

File: src/machine/machine.go (hunk 5 above)
Comment (deadprogram) [reply src/machine/machine.go]: Having the appropriate `Configure()` for PWM pins vs. GPIO pins seems to me like a nicer API. We can still avoid the interface for the moment.

File: src/examples/pwm/pwm.go (hunk 7 above)
Comment (deadprogram) [reply src/examples/pwm/pwm.go]: Dnoe,

**Comment by deadprogram** [comment]
Hi @aykevl pretty sure I have incorporated all your feedback on this PR. Please take another look if you can. Thanks.

File: src/machine/machine.go (hunk 5 above)
Comment (aykevl) [reply src/machine/machine.go]: Ah, right. `Configure()` (and `Set()`). Then I don't see an obvious solution. Let's keep it as is for now.

File: tools/gen-device-avr.py [hunk 8]
```diff
@@ -183,10 +190,15 @@ def writeGo(outdir, device):
...
-                out.write('\n')
+                name = bitfield['name']
+                value = bitfield['value']
+                bitshift = getShift(value)
+                for c in range(bin(value >> bitshift).count("1")):
+                    v = hex(1 << (bitshift + c))
+                    out.write('\t{0}{1} = {2}'.format(name, c, v))
+                    if bitfield['description']:
+                        out.write(' // {description}'.format(**bitfield))
+                    out.write('\n')
```
Comment (aykevl) [inline tools/gen-device-avr.py]: I think it's best to keep the old behavior when there is only one bit in the bitfield, so that src/runtime/runtime_avr.go doesn't have to be modified. This is more consistent with the AVR header files which don't append the suffix in that case. An easy way to check how many bits there are in a number is using `'{:08b}'.format(value).count('1')`.

Regarding the bit-suffix part: the  for loop can be simplified to:
```python
n = 0
for i in range(8):
    if (value >> i) & 1 == 0: continue
    out.write('\t{}{} = 0x{:x}'.format(name, n, 1 << i))
    if bitfield['description']:
        out.write(' // {description}'.format(**bitfield))
    n += 1
```
This doesn't need the `getShift()` function.

Comment (deadprogram) [reply tools/gen-device-avr.py]: Seems reasonable, changes made. 

Note that `UCSR0C_UCSZ00` actually did have 2 values, so that line in `runtime_avr.go` had to be changed to be consistent with the new naming.

**Comment by deadprogram** [comment]
This should now be ready for merge, please take a look. Once it lands I will start on the ADC.

File: src/runtime/runtime_avr.go [hunk 9]
```diff
@@ -37,7 +37,7 @@ func initUART() {
 	*avr.UBRR0H = 0
 	*avr.UBRR0L = 8
 	*avr.UCSR0B = avr.UCSR0B_RXEN0 | avr.UCSR0B_TXEN0 // enable RX and TX
-	*avr.UCSR0C = avr.UCSR0C_UCSZ0                    // 8-bits data
+	*avr.UCSR0C = avr.UCSR0C_UCSZ00                   // 8-bits data
```
Comment (aykevl) [inline src/runtime/runtime_avr.go]: This breaks UART: I had to change it to `*avr.UCSR0C = avr.UCSR0C_UCSZ01 | avr.UCSR0C_UCSZ00` to get UART to work again.

File: src/machine/machine_avr.go [hunk 10]
```diff
@@ -58,3 +58,68 @@ func (p GPIO) Get() bool {
...
+	if pwm.Pin < 8 {
+		*avr.DDRD |= 1 << pwm.Pin
+	} else {
+		*avr.DDRB |= 1 << (pwm.Pin - 8)
+	}
+}
+
+// Set turns on the duty cycle for a PWM pin using the provided value. On the AVR this is normally a
+// 8-bit value ranging from 0 to 255.
+func (pwm PWM) Set(value uint16) {
```
Comment (aykevl) [inline src/machine/machine_avr.go]: I see you have switched to 16-bit numbers for the duty cycle, but that wasn't entirely as I intended.
See [here](https://github.com/aykevl/tinygo/wiki/Machine-API) for details on my proposal (comments welcome). To be concrete, my idea for this PR was to use `uint8` everywhere as it's the fastest on AVR, but only convert to the full 16-bit range across the `PWM.Set()` boundary (using `.Set(uint16(value)*256)`). This way the PWM implementation can support full 16-bit PWM when available and fall back to 8-bit PWM (after dividing by 256) on hardware that doesn't have this precision. Doing it this way makes sure the caller can work without having to know how many bits the PWM has.

This means that on the AVR, the 8 bits PWM would be set using e.g. `OCR2B = avr.RegValue(value >> 8)` and 16-bits PWM can use something like `OCR1AL = avr.RegValue(value); OCR1AH = avr.RegValue(value >> 8);` for transparent 16-bit operation.

Sorry to be nitpicking here again, but I would really like to get the design right first (I missed it yesterday). If you have a different idea of how this should be implemented, please let me know.

File: src/runtime/runtime_avr.go (hunk 9 above)
Comment (deadprogram) [reply src/runtime/runtime_avr.go]: OK will add a commit to correct this.

File: src/machine/machine_avr.go (hunk 10 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: Took me a moment to follow what you were saying here, but then I noticed the wiki page. Did not realize that was there.

The only question I have is, is it better to have a single 16-bit scale that Arduino programmers have to write code that is different then the 0-254 they expect? My intent was to try to use the scaling that fits the docs for the board you are working with.

That is the reason I implemented it in this way, assuming that many people will simply enter the 8-bit value when "porting" Arduino code to TinyGo. We can enter a GoDoc comment to inform people they should use the 16-bit scaled value to try to steer them towards the correct value for their specific hardware.

File: src/runtime/runtime_avr.go (hunk 9 above)
Comment (deadprogram) [reply src/runtime/runtime_avr.go]: Done.

File: src/machine/machine_avr.go (hunk 10 above)
Comment (aykevl) [reply src/machine/machine_avr.go]: Yeah I wasn't sure how to express it clearly but glad you got it.

"Arduino programmers expect it" is not a very convincing argument to me, but googling PWM a bit more I see a different issue (obvious in hindsight): most MCUs are physically not able to reach a higher frequency without sacrifices on the frequency (because the clock speed is usually around 16MHz or not much higher). So maybe it's best to simply use 8 bits values for now and add `Set16()` and `SetFrequency()` (or something like that) later on for programs that require more precision.

In other words: I think simply using `uint8` is best after all. I can change that for you when merging (it's a fairly trivial change) or you can add a new commit. Both are fine to me.

Comment (deadprogram) [reply src/machine/machine_avr.go]: Well, I think the nrf ARM devices use 16-bit values, no? In which case we will have this same question immediately on the next platform. Seems like we should do the uint16 value as you first proposed.

Comment (aykevl) [reply src/machine/machine_avr.go]: I've looked into it and it seems that it uses 16-bit values. But the PWM there is very different from the one on the AVR (it even has DMA!). For example, it allows to set a custom 'top' value, meaning the max value that is supported (the dutycycle is then `duty/top*100%`). Additionally, there is a divisor based on the clock frequency (where the clock frequency is fixed to 16MHz), so you can get 16MHz, 8MHz, 4MHz, etc. Oh, and a single PWM module controls multiple pins, so it isn't possible to set the top value different for each pin.

Still, the issue with frequency vs accuracy remains, and is fairly fundamental (limited by clock speed). A possible implementation for nrf would be fixing the divisor to 16MHz, the top to 255, and control the PWM with the duty cycle (called a channel on the nrf) to get an equivalent implementation as for the Arduino. Programs that need more flexibility can use the registers directly or use a different package.

In other words, I think using `uint8` is fine for nrf.

Comment (deadprogram) [reply src/machine/machine_avr.go]: In that case, do you mind adding another commit to the PR to make it how you think it should be? I can test on my LED setup before merge if that is helpful, otherwise just land it!

Comment (aykevl) [reply src/machine/machine_avr.go]: I have made two implementations, one using `uint16` as I originally intended and one using `uint8` like you wrote originally: https://github.com/aykevl/tinygo/commit/56d843515fd726881115dcf6049bfd64155a5061 and https://github.com/aykevl/tinygo/commit/f2244539d4fd9e4e0e3c2bf069c9e6cfea1020a7.

They compile down to the exact same code, so when used like this it's a zero-cost abstraction to use 16-bit values. However, using `uint16` in the color calculations increases code size a bit as the calculations are done in 16 bits instead of 8.
Both implementations are fine by me, by now I'm not entirely sure what the best API is...
Do you have a preference?

> I can test on my LED setup before merge if that is helpful, otherwise just land it!

If you want, that would be useful (just to be sure).

Comment (deadprogram) [reply src/machine/machine_avr.go]: I vote for 16 bit duty cycle, since that will let us stick with that for pretty much all platforms in the immediate future.

File: src/examples/pwm/pwm.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+func cycleColor(color uint16) uint16 {
+	if color < 10 {
+		return color + 1
+	}
+	if color > 30 && color < 200 {
+		return color + 10
+	}
+	if color > 230 {
+		return 0
+	}
```
Comment (aykevl) [inline src/examples/pwm/pwm.go]: I don't think this function works correctly: it causes all colors to be in the range 0..10. Is that the intention?
Otherwise, maybe it should be like this? It slowly builds up from 0 to 10 and then jumps in bigger steps until 200 to start again at 0.
```go
func cycleColor(color uint8) uint8 {
        if color < 10 {
                return color + 1
        } else if color < 200 {
                return color + 10
        } else {
                return 0
        }
}
```
Discovered while testing the code.

Comment (deadprogram) [reply src/examples/pwm/pwm.go]: It is a bit ugly but it does test values greater than 10. This can be visually determined when looking at the LED.

Actually I was going to replace with a specific list of RGB colors to make it easier to spot check, but it does send several RGB colors to exercise various PWM duty cycle values, which was the main point.

Comment (aykevl) [reply src/examples/pwm/pwm.go]: I printed the numbers and for me it certainly never reaches beyond 10 and the LED stays rather dim. The LED gets a whole lot brighter with the above code.

Also, consider what happens when `cycleColor()` gets the value 10:

```go
	if color < 10 { // if not taken: 10 isn't smaller than 10
		return color + 1
	}
	if color > 30 && color < 200 { // if not taken: 10 isn't bigger than 30
		return color + 10
	}
	if color > 230 { // if not taken: 10 isn't bigger than 230
		return 0
	}
	// cycleColor(10) will return 0
	return 0
```

The same result can be obtained with input values 20 and 30 (for `gc` and `bc`) - these also return 0 restarting from the beginning.

Comment (deadprogram) [reply src/examples/pwm/pwm.go]: Sure, sounds good. Please change to your proposes function, thanks.

File: src/machine/machine_avr.go (hunk 10 above)
Comment (deadprogram) [reply src/machine/machine_avr.go]: After sleeping on it, I think I still prefer the `unint16` approach you defined in the wiki doc.

**Comment by aykevl** [comment]
Merged in fc0ff3a987ca227ce7fa7d1f1264ca6af010d1c6. Thank you for all the work!
